### DISCUSSION TITLE
No-cheating mode

### DISCUSSION THREAD

**Discussion by utaal** [verus-lang/verus#1292] [Feature requests]
Suggested by @parno.

As discussed in the Verus meeting on October 1st 2024 with @parno, @Chris-Hawblitzel, @jonhnet, @tjhance, @utaal, @jaybosamiya.

Chris suggests that a flag `--no-cheating` should be command line, not an attribute like `#[deny(cheating)]`.
@utaal, @Chris-Hawblitzel, @jaybosamiya: for crafting a specific policy for `#[verifier::external_body]` we may want rust-alike `forbid`, `deny`, `allow`.

@jonhnet, @parno: consider adding a list of files that allow `#[allow(cheating)]`. Tool enforces that are transitively closed.

**Comment by achreto** [comment]
With regards to `unsafe` usages in Rust, there are some tools that provides a nice audit list. 
For example: https://github.com/geiger-rs/cargo-geiger

When building a verified application, then I would like to know whether some of my dependencies are using assumes/external body. Having a tool like cargo-geiger for "cheats" that produces a report or so that shows 1) what is trusted and 2) the usages of assumes/external_body

**Comment by utaal** [comment]
@Chris-Hawblitzel: on a list of modules at the root of the crate, you can `forbid(external_body)` on each module, except those that are trusted.
@utaal: in fact that may be *the* way to provide the list of trusted modules: those that **do not** have `forbid(external_body)`.
@jaybosamiya: let's not use filenames.

@Chris-Hawblitzel, @utaal: instead let's have a crate-root-level mechanism to mark allowed `external_body`:
`#![forbid_all(external_body)]` at the root of the crate;
`#[unforbid(external_body)]` on the `mod name;` lines in the crate root.

@Chris-Hawblitzel: Verus can enforce a policy that `unforbid`-en modules are constrained to e.g. which file name to use.

**Reply by parno** [reply]
To expand on (my interpretation of) Chris's suggestion: We can use familiar-looking Rust syntax (e.g., the various [lint levels](https://doc.rust-lang.org/rustc/lints/levels.html)) to make the notation accessible, but Verus can impose additional policy on top of this when run with `--no-cheating`.  The policy can, for example, enforce:
1. The crate root must start with `#![deny(verus::assumptions)]`.
2.  The `#[allow(verus::assumptions)]` annotation can only appear in the crate root, and only on the `mod name` lines.
3. An `#[allow(verus::assumptions)]` module must correspond to a file-level module (to satisfy Jon's easy-to-audit criteria)
4. Modules on the `allow` list must be transitively closed; they can't use any non-`allow` modules

**Comment by parno** [comment]
Note that we also discussed connecting `--no-cheating` to exec-mode [termination checking](https://github.com/verus-lang/verus/discussions/1176#discussioncomment-9875316).

There's also some older but possibly relevant discussion w.r.t. [trusted/untrusted splits](https://github.com/verus-lang/verus/discussions/112).

**Reply by tjhance** [reply]
What should `--no-cheating` imply the enforcement of termination-checking?

**Reply by parno** [reply]
I think in most contexts, `--no-cheating` implies all post-conditions have been verified; i.e., it indicates a rejection of weakest liberal preconditions.  For example, if I'm grading a class assignment or evaluating an LLM on whether a proof has been completed, I don't want to accept a program that is incorrect but verifies due to the presence of an infinite loop.

**Reply by tjhance** [reply]
What are these other contexts? Are there other tools that implement a no-cheating flag?

Non-terminating programs are sometimes desirable, and if you want to prove termination you should be able to put it on the function specification. I think `--no-cheating` should have a single coherent aim: to remove threats to the validity of your specification, which it can do by disallowing external_body, assume, etc. But if you require termination checking ... suddenly it has an opinion on what your specification _is_, which is weird? What if I want to use no-cheating for some infinite-looping code?

**Reply by parno** [reply]
> What are these other contexts? 
1. Teaching
2. LLMs
3. I think most of the projects listed on the [Verus projects](https://verus-lang.github.io/verus/publications-and-projects/) page, plus Owl.  IronFleet has one top-level infinite loop, but that's in a trusted file.

>  Are there other tools that implement a no-cheating flag?

Dafny has one (`/noCheating`), but some of it may now be on by default, since there's also an option `--allow-axioms`, which defaults to `false` (or this may be due to their migration from the old approach to command-line arguments to the new verb-based approach).  Dafny already insists on termination checking by default, however, unless you systematically opt out via `decreases *` (which is then visible as part of your top-level theorem), so it doesn't need to include termination in `/noCheating`.

At a higher level, I agree there are valid reasons to have non-terminating code, and that you might want to use `--no-cheating` on such code.  OTOH, I hypothesize that most users would expect `--no-cheating` to rule out a "successful" verification that arises due to an overlooked non-terminating loop, and would be sad to belatedly discover that problem.  Maybe we can find a way to eliminate that surprise via the right set of options and/or careful documentation in the right places.

**Reply by zeldovich** [reply]
Bryan's desire for termination checking to catch students that get out of proving a postcondition by introducing an infinite loop feels pretty different from wanting to know the function will actually return (which means also proving things like there's no deadlock in calls to acquire locks).

Would you imagine that `--no-cheating` should prevent the use of locks?  That seems pretty severe for the kinds of use cases I would imagine (e.g., compiling all `*_v.rs` files with `--no-cheating` to make sure there's no axioms being introduced by accident there).

**Reply by tjhance** [reply]
It feels really strange to me that a tool whose primary purpose is to disallow `assume` statements would also disallow a class of perfectly legitimate specifications, just because they're slightly rare. It seems like a conflation of different issues to me, one that could lead to people wrongly concluding that nonterminating code is illegitimate, or that Verus doesn't support it. I also think most of your issues could be resolved by making termination-checking the default (so that you have to write a very conspicuous `must_terminate false` or something in the specification if you want to allow non-terminating code).

**Reply by parno** [reply]
> Would you imagine that --no-cheating should prevent the use of locks? 

No, I think our current plan for enforcing exec mode termination is to only target it at single-threaded code.  As you say, enforcing termination for concurrent code quickly becomes quite tricky.  Even in single-threaded code, you have to make assumptions about the APIs you call (including the Rust standard library).  Part of the challenge here is that I'm not sure what the exec-mode termination checking will look like, or what options it will allow.  Perhaps the connection to `no-cheating` will be clearer once the PR for that arrives.

> It feels really strange to me that a tool whose primary purpose is to disallow assume statements would also disallow a class of perfectly legitimate specifications

I think of it more broadly as a way of ensuring that when I look at a top-level method's specification, I know that if I satisfy the preconditions, I will get a result that satisfies the postcondition.  From a CI perspective, I also want to know that everything that method depends on has been verified, not that someone might have forgotten to write `i = i + 1` somewhere in the code base and as a result we haven't actually been verifying some non-trivial swath of the code.

> I also think most of your issues could be resolved by making termination-checking the default (so that you have to write a very conspicuous must_terminate false or something in the specification if you want to allow non-terminating code).

Depending on how the exec-termination checking is implemented, this could certainly be a way to approach it.  I do think it's important that must be visible at the top-level, not something you can "hide" deep in the implementation.

**Reply by tjhance** [reply]
> From a CI perspective, I also want to know that everything that method depends on has been verified, not that someone might have forgotten to write i = i + 1 somewhere in the code base and as a result we haven't actually been verifying some non-trivial swath of the code.

Maybe there's a little confusion here ... I'm imagining that when we have official termination-checking support it will, when enabled for a particular function, actually _verify_ termination, i.e., it will check all the dependencies terminate, and so on.

**Reply by parno** [reply]
There's been discussion of allowing various granularities of opting-in/opting-out of termination checking, so I'd like to see what state that ends up in.

**Comment by utaal** [comment]
Implemented as https://github.com/verus-lang/verus/pull/1499

Should be complemented by enforced termination checking (wip by @apribiso).

**Comment by nickgian** [comment]
Reviving this discussion, it would be nice to have a `--no-cheating` mode that is acting as an audit mechanism rather than an enforcement one. 

When a program verifies using assumptions about external libraries, vstd, or admitted facts in my own code, I want to inspect exactly what was trusted without causing verification to fail.

`--no-cheating` gets close, but has two limitations for this use case:
1. Disallowed assumptions make the invocation fail, it does not report “verification succeeded, but the no-cheating policy failed.”
2. The JSON diagnostics do not have a stable machine-readable identifier for "trust" errors. You get something like:
```
  {
     "message": "external_body/assume_specification not allowed with --no-cheating",
     "code": null,
     "level": "error"
  }
```
One would need to know what message they need to parse to identify the issue and keep up with any Verus changes/additions.

I see two alternatives:
- Some form of lint-like severity levels as discussed in this thread + diagnostic codes that make it easier to programmatically identify the issue.
- A dedicated TCB inventory that is emitted via a flag like `--log tcb` that is purely informational and does not interact with rustc diagnostics. Both can be derived from the same code essentially.

The latter creates a parallel mechanism and possible confusion. On the other hand it feels a bit easier to audit the TCB with a targeted tool like this, rather than trying to describe what is allowed via the more expressive Rust lints. Especially in the context of agentic systems, it's hard to trust an agent won't add an `allow(verus::assumption)` somewhere I did not want it to (so now I have to audit that instead :D).
